lib/Dialect/XTenNN/IR/XTenNNOps.cpp

```diff
@@ -0,0 +1,183 @@
+///
+/// See parseCaptures() for more details.
+static void printCaptures(OpAsmPrinter &p, ValueRange args, ValueRange srcs) {
+  auto srcIt = srcs.begin();
+  p << '(';
+  llvm::interleaveComma(args, p, [&](auto arg) {
+    assert(srcIt != srcs.end());
+    printCapture(p, arg, *srcIt++);
+  });
+  p << ')';
+}
+
+/// Parses a trivial EnclaveOp.
+///
+/// Format:
+///     capture-list [ attr-dict-with-keyword ] region [ `->` type-list ]
+static ParseResult parseEnclaveOp(OpAsmParser &p, OperationState &result) {
+  // `(` captures `)`
+  SmallVector<OpAsmParser::Argument> args;
+  if (parseCaptures(p, args, result.operands))
+    return failure();
+
+  // [ attr-dict-with-keyword ]
+  if (p.parseOptionalAttrDictWithKeyword(result.attributes))
+    return failure();
+
+  // `{` ... `}`
+  if (p.parseRegion(*result.addRegion(), args, true))
+    return failure();
+
+  // [ `->` type-list ]
+  if (succeeded(p.parseOptionalArrow())) {
+    if (p.parseTypeList(result.types))
+      return failure();
+  }
+
+  return success();
+}
+
+/// Prints a trivial EnclaveOp.
+///
+/// See parseEnclaveOp() for more details.
+static void printEnclaveOp(OpAsmPrinter &p, EnclaveOp op) {
+  p << ' ';
+  printCaptures(p, op.getEnclaveBody().getArguments(), op.getCaptures());
+  p << ' ';
+
+  p.printOptionalAttrDictWithKeyword(op->getAttrs());
+  if (!op->getAttrs().empty())
+    p << ' ';
+
+  p.printRegion(*op.getEnclaveBody().getParent(), false);
+
+  if (op->getNumResults() > 0) {
+    p << " -> ";
+    interleaveComma(op->getResultTypes(), p);
+  };
+}
+
+#define GET_OP_CLASSES
+#include "xten/Dialect/XTenNN/IR/XTenNNOps.cpp.inc"
+
+//===----------------------------------------------------------------------===//
+// SubgraphOp
+//===----------------------------------------------------------------------===//
+
+ParseResult SubgraphOp::parse(OpAsmParser &p, OperationState &result) {
+  return parseEnclaveOp(p, result);
+}
+
+void SubgraphOp::print(OpAsmPrinter &p) {
+  printEnclaveOp(p, *this);
+}
+
+LogicalResult SubgraphOp::verify() {
+  // The number of captures must match the number of block arguments
+  if (this->getCaptures().size() != this->getEnclaveBody().getNumArguments()) {
+    return this->emitOpError()
+           << "number of operands (" << this->getCaptures().size()
+           << ") does not match number of block arguments ("
+           << this->getEnclaveBody().getNumArguments() << ")";
+  }
+
+  // The type of the arguments must match the types of the block arguments
+  for (auto [idx, argType] :
+       enumerate(this->getEnclaveBody().getArgumentTypes())) {
+    if (this->getCapture(idx).getType() != argType) {
+      return this->emitOpError()
+             << "type of operand #" << idx << " ("
+             << this->getCapture(idx).getType()
+             << ") does not match enclave argument type (" << argType << ")";
+    }
+  }
+  return enclave_interface_defaults::verify(*this);
```

Comment:
This is not necessary. The interface verifier is run by the TableGen generated `verifyInvariants` method, which processes all traits (including interface) and then calls this method. You should just return success here.

Reply:
Fixed